pkg/controller/endpointslice/endpointslice_tracker.go

```diff
@@ -73,33 +79,42 @@ func (est *endpointSliceTracker) Update(endpointSlice *discovery.EndpointSlice)
 	est.lock.Lock()
 	defer est.lock.Unlock()
 
-	rrv := est.relatedResourceVersions(endpointSlice)
+	rrv, ok := est.relatedResourceVersions(endpointSlice)
+	if !ok {
+		rrv = endpointSliceResourceVersions{}
+		est.resourceVersionsByService[getServiceNN(endpointSlice)] = rrv
+	}
 	rrv[endpointSlice.Name] = endpointSlice.ResourceVersion
 }
 
+// DeleteService removes the set of resource versions tracked for the Service.
+func (est *endpointSliceTracker) DeleteService(namespace, name string) {
+	est.lock.Lock()
+	defer est.lock.Unlock()
+
+	serviceNN := types.NamespacedName{Name: name, Namespace: namespace}
+	delete(est.resourceVersionsByService, serviceNN)
+}
+
 // Delete removes the resource version in this endpointSliceTracker for the
 // provided EndpointSlice.
 func (est *endpointSliceTracker) Delete(endpointSlice *discovery.EndpointSlice) {
 	est.lock.Lock()
 	defer est.lock.Unlock()
 
-	rrv := est.relatedResourceVersions(endpointSlice)
-	delete(rrv, endpointSlice.Name)
+	rrv, ok := est.relatedResourceVersions(endpointSlice)
+	if ok {
+		delete(rrv, endpointSlice.Name)
+	}
 }
 
 // relatedResourceVersions returns the set of resource versions tracked for the
-// Service corresponding to the provided EndpointSlice. If no resource versions
-// are currently tracked for this service, an empty set is initialized.
-func (est *endpointSliceTracker) relatedResourceVersions(endpointSlice *discovery.EndpointSlice) endpointSliceResourceVersions {
+// Service corresponding to the provided EndpointSlice, and a bool to indicate
+// if it exists.
+func (est *endpointSliceTracker) relatedResourceVersions(endpointSlice *discovery.EndpointSlice) (endpointSliceResourceVersions, bool) {
 	serviceNN := getServiceNN(endpointSlice)
 	vers, ok := est.resourceVersionsByService[serviceNN]
-
-	if !ok {
-		vers = endpointSliceResourceVersions{}
-		est.resourceVersionsByService[serviceNN] = vers
-	}
-
-	return vers
+	return vers, ok
```

Comment:
Maybe just return on the line above - `return est.resourceVersionsByService[serviceNN]`

Reply:
It's a single value when returning the map value, would cause compilation error "not enough arguments to return".

Reply:
Woops, somehow I hadn't run into that before, thanks for the explanation!